### PR TITLE
Add oci-image support for WhiteSource

### PR DESCRIPTION
**Pull request by zkdev**
**What this PR does / why we need it**:
This PR adds support for WhiteSource oci-image scans.
Each oci-layer is pulled into a seperate tar file, and a top level tar is sent to the WhiteSource API.
The config blob is ignored.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: oci/__init__.py [hunk 1]
```diff
@@ -224,17 +224,23 @@ def image_layers_as_tarfile_generator(
...
     returns a generator yielding a tar-archive with the passed oci-image's layer-blobs as
     members. This is somewhat similar to the result of a `docker save` with the notable difference
     that the cfg-blob is discarded.
     This function is useful to e.g. upload file system contents of an oci-container-image to some
     scanning-tool (provided it supports the extraction of tar-archives)
+    If include_config_blob is set to False the config blob will be ignored.
     '''
     manifest = oci_client.manifest(image_reference=image_reference)
     offset = 0
     for blob in manifest.blobs():
```
Comment (ccwienk) [inline oci/__init__.py:239]: rather mv the `if` upwards, like so:

```
blobs = manifest.blobs() if include_config_blob else manifest.layers
```

File: whitesource/component.py [hunk 2]
```diff
@@ -23,23 +23,24 @@ def get_scan_artifacts_from_components(
...
-            ws_hint = _get_ws_label_from_source(source)
-
-            if not ws_hint or ws_hint.policy is dso.labels.ScanPolicy.SCAN:
-                yield dso.model.ScanArtifact(
-                    access=source.access,
-                    label=ws_hint,
-                    name=f'{component.name}_{source.identity(component.sources)}'
-                )
-            elif ws_hint.policy is dso.labels.ScanPolicy.SKIP:
+        for source in component.sources + component.resources:
```
Comment (ccwienk) [inline whitesource/component.py:26]: naming-nit: we call the "base-class" of sources and resources `artifacts` in CNUDIE-CD

File: whitesource/component.py [hunk 3]
```diff
@@ -23,23 +23,24 @@ def get_scan_artifacts_from_components(
...
+            if not source.access:
                 continue
+            if source.access.type in (cm.AccessType.GITHUB, cm.AccessType.OCI_REGISTRY):
+                ws_hint = _get_ws_label_from_source(source)
```
Comment (ccwienk) [inline whitesource/component.py:30]: -> from_artifact

File: whitesource/component.py [hunk 4]
```diff
@@ -23,23 +23,24 @@ def get_scan_artifacts_from_components(
...
-                yield dso.model.ScanArtifact(
-                    access=source.access,
-                    label=ws_hint,
-                    name=f'{component.name}_{source.identity(component.sources)}'
-                )
-            elif ws_hint.policy is dso.labels.ScanPolicy.SKIP:
+        for source in component.sources + component.resources:
+            if not source.access:
                 continue
+            if source.access.type in (cm.AccessType.GITHUB, cm.AccessType.OCI_REGISTRY):
```
Comment (ccwienk) [inline whitesource/component.py:29]: I would as a general pattern prefer to invert the logic
--> skip / continue _unless_ we are interested in the artifact. thus indention for the actual "good" case can be reduced by one

File: whitesource/util.py [hunk 5]
```diff
@@ -203,44 +207,55 @@ def scan_artifact_with_white_src(
...
+                repository=scan_artifact.access.repository_name(),
+            )
+            # guess git-ref for the given version
+            commit_hash = product.util.guess_commit_from_source(
+                artifact_name=scan_artifact.name,
+                commit_hash=scan_artifact.access.commit,
+                ref=scan_artifact.access.ref,
+                github_repo=github_repo,
+            )
+            exclude_regexes = ''
```
Comment (ccwienk) [inline whitesource/util.py:226]: rather use either `None` or empty tuple?

**Review by ccwienk — CHANGES_REQUESTED** [review]
actually, I would prefer to only have one "if" that differentiates between the access-type

If I properly understand it, the contents to be uploaded seem (currently) to always be written into a tempfile anyhow.

so either 

a) create the file upfront (from either datasource) and pass it to the uploader
b) pass a generator / filelike-object (abstracting the actual datasource) to the uploader

probably a) will be the variant requiring less implementation efforts, considering you need to know the Content-Length upfront

File: oci/__init__.py [hunk 6]
```diff
@@ -224,17 +224,23 @@ def image_layers_as_tarfile_generator(
...
+        logger.debug(f'getting blob {blob.digest}')
+        if not include_config_blob:
+            if blob == manifest.config:
+                continue
```
Comment (AndreasBurger) [inline oci/__init__.py:243]: consider 
```suggestion
        if not include_config_blob and blob == manifest.config:
                continue
```
up to your preferences, of course

File: whitesource/util.py [hunk 7]
```diff
@@ -4,17 +4,21 @@
 import logging
 import tempfile
 import typing
+import os.path
```
Comment (AndreasBurger) [inline whitesource/util.py:7]: nit: sorting

File: whitesource/util.py [hunk 8]
```diff
@@ -203,44 +207,55 @@ def scan_artifact_with_white_src(
...
-        artifact_name=scan_artifact.name,
-        commit_hash=scan_artifact.access.commit,
-        ref=scan_artifact.access.ref,
-        github_repo=github_repo,
-    )
-    exclude_regexes = ''
-    include_regexes = ''
+    logger.debug('init scan')
+    with tempfile.NamedTemporaryFile() as tmp_file:
+        if scan_artifact.access.type == cm.AccessType.GITHUB:
```
Comment (AndreasBurger) [inline whitesource/util.py:212]:
```suggestion
        if scan_artifact.access.type is cm.AccessType.GITHUB:
```

**Review by ccwienk — APPROVED** [review]

File: whitesource/util.py [hunk 9]
```diff
@@ -203,44 +207,55 @@ def scan_artifact_with_white_src(
...
+                include_regexes=include_regexes
+            )
+            whitesource.component.download_component(
+                logger=logger,
+                github_repo=github_repo,
+                path_filter_func=path_filter_func,
+                ref=commit_hash,
+                target=tmp_file,
+            )
+        elif scan_artifact.access.type == cm.AccessType.OCI_REGISTRY:
```
Comment (AndreasBurger) [inline whitesource/util.py:245]:
```suggestion
        elif scan_artifact.access.type is cm.AccessType.OCI_REGISTRY:
```

**Review by AndreasBurger — APPROVED** [review]

**Pull request merged**